app/forms/shared/volunteering_role_form.rb

```diff
@@ -0,0 +1,121 @@
+module Shared
+  class VolunteeringRoleForm
+    include ActiveModel::Model
+    include DateValidationHelper
+
+    attr_accessor :id,
+                  :role,
+                  :organisation,
+                  :details,
+                  :working_with_children,
+                  :start_date_day,
+                  :start_date_month,
+                  :start_date_year,
+                  :start_date_unknown,
+                  :currently_working,
+                  :end_date_day,
+                  :end_date_month,
+                  :end_date_year,
+                  :end_date_unknown
+
+    validates :role,
+              :organisation,
+              :working_with_children,
+              presence: true
+
+    validates :role, :organisation, length: { maximum: 60 }
+    validates :start_date, date: { presence: true, future: true, month_and_year: true, before: :end_date }
+
+    validates :start_date_unknown, inclusion: { in: %w[true false] }
+    validates :end_date_unknown, inclusion: { in: %w[true false] }
+    validates :currently_working, presence: true
+    validates :currently_working, inclusion: { in: %w[true false] }, if: -> { currently_working.present? }
+    validates :end_date, date: { presence: true, future: true, month_and_year: true }, if: -> { currently_working == 'false' }
+
+    validates :end_date, date: { future: true, month_and_year: true }, unless: :dont_validate_end_date
+
+    validates :details, presence: true, word_count: { maximum: 150 }
+
+    class << self
+      def build_all_from_application(application_form)
+        application_form
+          .application_volunteering_experiences
+          .order('created_at DESC', :id).map do |volunteering_role|
+          new_volunteering_role_form(volunteering_role)
+        end
+      end
+
+      def build_from_experience(experience)
+        new_volunteering_role_form(experience)
+      end
+
+    private
+
+      def new_volunteering_role_form(volunteering_role)
+        new(
+          id: volunteering_role.id,
+          role: volunteering_role.role,
+          organisation: volunteering_role.organisation,
+          details: volunteering_role.details,
+          working_with_children: volunteering_role.working_with_children,
+          start_date_month: volunteering_role.start_date.month,
+          start_date_year: volunteering_role.start_date.year,
+          start_date_unknown: volunteering_role.start_date_unknown.to_s,
+          end_date_month: volunteering_role.end_date&.month || '',
+          end_date_year: volunteering_role.end_date&.year || '',
+          end_date_unknown: volunteering_role.end_date_unknown.to_s,
+          currently_working: volunteering_role.currently_working.to_s,
+        )
+      end
+    end
+
+    def save(application_form)
+      return false unless valid?
+
+      application_form.application_volunteering_experiences.create!(map_attributes)
+      application_form.update!(volunteering_experience: true)
```

Comment:
Maybe wrapping this in a transaction is safer

Reply:
That code got copied across. Added the transaction [here](https://github.com/DFE-Digital/apply-for-teacher-training/pull/8713/files#diff-4bf3f35ac586d93e16c4f2ac873ae892225ddfd2b50606aaa63d74016bf609f4R75).